test/Microsoft.AspNetCore.SystemWebAdapters.CoreServices.Tests/SessionState/SessionIntegrationTests.cs

```diff
@@ -0,0 +1,195 @@
+public class SessionIntegrationTests
+{
+    [InlineData("/?override=disabled", "Session:null")]
+    [InlineData("/?override=readonly", "ReadOnly:True")]
+    [InlineData("/?override=required", "ReadOnly:False")]
+    [InlineData("/?override=default", "Session:null")]
+    [Theory]
+    public async Task TestSetSessionStateBehavior(string endpoint, string expected)
+    {
+        var actual = await GetAsync(endpoint);
+        Assert.Equal(expected, actual);
+    }
+
+    [InlineData("/disabled", "Session:null")]
+    [InlineData("/readonly", "ReadOnly:True")]
+    [InlineData("/required", "ReadOnly:False")]
+    [InlineData("/default", "Session:null")]
+    [Theory]
+    public async Task TestSessionAttribute(string endpoint, string expected)
+    {
+
+        var actual = await GetAsync(endpoint);
+        Assert.Equal(expected, actual);
+    }
+
+    [InlineData("/disabled?override=required", "ReadOnly:False")]
+    [InlineData("/disabled?override=readonly", "ReadOnly:True")]
+    [InlineData("/readonly?override=required", "ReadOnly:False")]
+    [InlineData("/default?override=disabled", "Session:null")]
+    [Theory]
+    public async Task TestOverrideSessionStateBehavior(string endpoint, string expected)
+    {
+        var actual = await GetAsync(endpoint);
+        Assert.Equal(expected, actual);
+    }
+
+    private static async Task<string> GetAsync(string endpoint)
+    {
+        using var host = await new HostBuilder()
+          .ConfigureWebHost(webBuilder =>
+          {
+              webBuilder
+                  .UseTestServer(options =>
+                  {
+                      options.AllowSynchronousIO = true;
+                  })
+                  .ConfigureServices(services =>
+                  {
+                      services.AddRouting();
+                      services.AddControllers();
+                      services.AddSystemWebAdapters()
+                        .WrapAspNetCoreSession();
+                      services.AddDistributedMemoryCache();
+                  })
+                  .Configure(app =>
+                  {
+                      app.UseRouting();
+                      app.Use((ctx, next) =>
+                      {
+                          SetOverrideSessionState(ctx);
+                          return next(ctx);
+                      });
+                      app.UseSession();
+                      app.UseSystemWebAdapters();
+                      app.UseEndpoints(endpoints =>
+                      {
+                          endpoints.MapControllers();
+
+                      });
+                  });
+          })
+          .StartAsync();
+
+        var uri = new Uri(endpoint, UriKind.Relative);
+
+        try
+        {
+            return await host.GetTestClient().GetStringAsync(uri).ConfigureAwait(false);
+        }
+        finally
+        {
+            await host.StopAsync();
+        }
+    }
+
+    private static void SetOverrideSessionState(HttpContext context)
```

Comment:
can you inline this method to the test? It'll make it easier to understand what's happening